File: petshop_app/lib/models/product.dart
class Product {
  final int id;
  final int sellerId;
  final String title;
  final String description;
  final int categoryId;
  final List<String> images;
  final String status;
  final ProductType type;
  final AuctionInfo? auctionInfo;
  final FixedInfo? fixedInfo;
  final String? location;
  final Seller? seller;
  final DateTime createdAt;
  final DateTime updatedAt;

  Product({
    required this.id,
    required this.sellerId,
    required this.title,
    required this.description,
    required this.categoryId,
    required this.images,
    required this.status,
    required this.type,
    this.auctionInfo,
    this.fixedInfo,
    this.location,
    this.seller,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    // 基于实际API响应格式解析
    final int auctionType = json['auction_type'] ?? 1;
    final ProductType productType =
        auctionType == 1 ? ProductType.auction : ProductType.fixed;

    // 构建AuctionInfo（如果是拍卖类型）
    AuctionInfo? auctionInfo;
    if (productType == ProductType.auction) {
      auctionInfo = AuctionInfo(
        startPrice:
            double.tryParse(json['starting_price']?.toString() ?? '0') ?? 0.0,
        currentPrice:
            double.tryParse(json['current_price']?.toString() ?? '0') ?? 0.0,
        bidIncrement: 10.0, // 默认值，API中没有这个字段
        startTime: json['auction_start_time'] != null
            ? DateTime.parse(json['auction_start_time'])
            : DateTime.now(),
        endTime: json['auction_end_time'] != null
            ? DateTime.parse(json['auction_end_time'])
            : DateTime.now().add(Duration(days: 7)),
        bidCount: json['bid_count'] ?? 0,
      );
    }

    // 构建FixedInfo（如果是一口价类型）
    FixedInfo? fixedInfo;
    if (productType == ProductType.fixed) {
      fixedInfo = FixedInfo(
        price: double.tryParse(json['buy_now_price']?.toString() ?? '0') ?? 0.0,
        stock: json['stock_quantity'] ?? 0,
        salesCount: 0, // API中没有这个字段
      );
    }

    return Product(
      id: json['id'] ?? 0,
      sellerId: json['seller_id'] is int
          ? json['seller_id']
          : (int.tryParse(json['seller_id']?.toString() ?? '0') ?? 0),
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      categoryId: json['category_id'] ?? 0,
      images: List<String>.from(json['images'] ?? []),
      status: (json['status'] ?? 1).toString(),
      type: productType,
      auctionInfo: auctionInfo,
      fixedInfo: fixedInfo,
      location: json['location'],
      seller: null, // seller信息需要单独接口获取
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'seller_id': sellerId,
      'title': title,
      'description': description,
      'category_id': categoryId,
      'images': images,
      'status': status,
      'type': type.toString(),
      'auction_info': auctionInfo?.toJson(),
      'fixed_info': fixedInfo?.toJson(),
      'location': location,
      'seller': seller?.toJson(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  // 获取当前价格
  double get currentPrice {
    if (type == ProductType.auction) {
      return auctionInfo?.currentPrice ?? 0;
    } else {
      return fixedInfo?.price ?? 0;
    }
  }

  // 获取主图
  String get mainImage {
    return images.isNotEmpty ? images.first : '';
  }

  // 是否正在拍卖中
  bool get isAuctionActive {
    if (type != ProductType.auction || auctionInfo == null) return false;
    final now = DateTime.now();
    return now.isAfter(auctionInfo!.startTime) &&
        now.isBefore(auctionInfo!.endTime);
  }

  // 获取剩余时间
  Duration? get timeLeft {
    if (type != ProductType.auction || auctionInfo == null) return null;
    final now = DateTime.now();
    if (now.isAfter(auctionInfo!.endTime)) return null;
    return auctionInfo!.endTime.difference(now);
  }
}

enum ProductType {
  auction,
  fixed;

  static ProductType fromString(String value) {
    switch (value.toLowerCase()) {
      case 'auction':
        return ProductType.auction;
      case 'fixed':
        return ProductType.fixed;
      default:
        return ProductType.auction;
    }
  }

  @override
  String toString() {
    switch (this) {
      case ProductType.auction:
        return 'auction';
      case ProductType.fixed:
        return 'fixed';
    }
  }
}

class AuctionInfo {
  final double startPrice;
  final double currentPrice;
  final double bidIncrement;
  final DateTime startTime;
  final DateTime endTime;
  final int bidCount;

  AuctionInfo({
    required this.startPrice,
    required this.currentPrice,
    required this.bidIncrement,
    required this.startTime,
    required this.endTime,
    required this.bidCount,
  });

  factory AuctionInfo.fromJson(Map<String, dynamic> json) {
    return AuctionInfo(
      startPrice: (json['start_price'] ?? 0).toDouble(),
      currentPrice: (json['current_price'] ?? 0).toDouble(),
      bidIncrement: (json['bid_increment'] ?? 0).toDouble(),
      startTime: DateTime.parse(
          json['start_time'] ?? DateTime.now().toIso8601String()),
      endTime:
          DateTime.parse(json['end_time'] ?? DateTime.now().toIso8601String()),
      bidCount: json['bid_count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'start_price': startPrice,
      'current_price': currentPrice,
      'bid_increment': bidIncrement,
      'start_time': startTime.toIso8601String(),
      'end_time': endTime.toIso8601String(),
      'bid_count': bidCount,
    };
  }
}

class FixedInfo {
  final double price;
  final int stock;
  final int salesCount;

  FixedInfo({
    required this.price,
    required this.stock,
    required this.salesCount,
  });

  factory FixedInfo.fromJson(Map<String, dynamic> json) {
    return FixedInfo(
      price: (json['price'] ?? 0).toDouble(),
      stock: json['stock'] ?? 0,
      salesCount: json['sales_count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'price': price,
      'stock': stock,
      'sales_count': salesCount,
    };
  }
}

class Seller {
  final int id;
  final String name;
  final String? avatar;
  final double rating;
  final int followerCount;

  Seller({
    required this.id,
    required this.name,
    this.avatar,
    required this.rating,
    required this.followerCount,
  });

  factory Seller.fromJson(Map<String, dynamic> json) {
    return Seller(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      avatar: json['avatar'],
      rating: (json['rating'] ?? 0).toDouble(),
      followerCount: json['follower_count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'avatar': avatar,
      'rating': rating,
      'follower_count': followerCount,
    };
  }
}

class Category {
  final int id;
  final String name;
  final int? parentId;
  final int sortOrder;
  final String? icon;
  final String status;

  Category({
    required this.id,
    required this.name,
    this.parentId,
    required this.sortOrder,
    this.icon,
    required this.status,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      parentId: json['parent_id'],
      sortOrder: json['sort_order'] ?? 0,
      icon: json['icon'],
      status: json['status'] ?? 'active',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'parent_id': parentId,
      'sort_order': sortOrder,
      'icon': icon,
      'status': status,
    };
  }
}

File: backend/app/models/product.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, DECIMAL, JSON
from sqlalchemy.sql import func
from app.core.database import Base

class Category(Base):
    __tablename__ = "categories"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    name = Column(String(50), nullable=False)
    parent_id = Column(Integer, default=0)
    icon_url = Column(String(500))
    sort_order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

class Product(Base):
    __tablename__ = "products"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    seller_id = Column(Integer, nullable=False, index=True)
    category_id = Column(Integer, nullable=False, index=True)
    title = Column(String(200), nullable=False)
    description = Column(Text)
    images = Column(JSON, comment="商品图片数组")
    starting_price = Column(DECIMAL(10, 2), nullable=False)
    current_price = Column(DECIMAL(10, 2), nullable=False)
    buy_now_price = Column(DECIMAL(10, 2), comment="一口价")
    auction_type = Column(Integer, default=1, comment="1:拍卖,2:一口价,3:混合")
    auction_start_time = Column(DateTime)
    auction_end_time = Column(DateTime)
    location = Column(String(100))
    shipping_fee = Column(DECIMAL(8, 2), default=0.00)
    is_free_shipping = Column(Boolean, default=False)
    condition_type = Column(Integer, default=1, comment="1:全新,2:二手,3:其他")
    stock_quantity = Column(Integer, default=1)
    view_count = Column(Integer, default=0)
    bid_count = Column(Integer, default=0)
    favorite_count = Column(Integer, default=0)
    status = Column(Integer, default=1, comment="1:待审核,2:拍卖中,3:已结束,4:已下架")
    is_featured = Column(Boolean, default=False)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

class Bid(Base):
    __tablename__ = "bids"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    product_id = Column(Integer, nullable=False, index=True)
    bidder_id = Column(Integer, nullable=False, index=True)
    bid_amount = Column(DECIMAL(10, 2), nullable=False)
    is_auto_bid = Column(Boolean, default=False)
    max_bid_amount = Column(DECIMAL(10, 2))
    status = Column(Integer, default=1, comment="1:有效,2:被超越,3:撤销")
    created_at = Column(DateTime, server_default=func.now())

class ProductImage(Base):
    __tablename__ = "product_images"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    product_id = Column(Integer, nullable=False, index=True)
    image_url = Column(String(500), nullable=False)
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime, server_default=func.now())

class ProductFavorite(Base):
    __tablename__ = "product_favorites"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    user_id = Column(Integer, nullable=False, index=True)
    product_id = Column(Integer, nullable=False, index=True)
    created_at = Column(DateTime, server_default=func.now())

class Shop(Base):
    __tablename__ = "shops"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    owner_id = Column(Integer, nullable=False, index=True)
    shop_name = Column(String(100), nullable=False)
    shop_logo = Column(String(500))
    description = Column(Text)
    business_license = Column(String(100))
    contact_phone = Column(String(20))
    address = Column(String(200))
    rating = Column(DECIMAL(3, 2), default=5.00)
    total_sales = Column(Integer, default=0)
    status = Column(Integer, default=1, comment="1:正常,2:暂停,3:关闭")
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

class LocalService(Base):
    __tablename__ = "local_services"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    provider_id = Column(Integer, nullable=False, index=True)
    service_type = Column(Integer, nullable=False, comment="1:上门服务,2:宠物交流,3:鱼缸造景")
    title = Column(String(200), nullable=False)
    description = Column(Text)
    images = Column(JSON)
    price = Column(DECIMAL(10, 2))
    location = Column(String(100))
    contact_info = Column(String(100))
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

class SpecialEvent(Base):
    __tablename__ = "special_events"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    title = Column(String(100), nullable=False)
    description = Column(Text)
    banner_image = Column(String(500))
    start_time = Column(DateTime, nullable=False)
    end_time = Column(DateTime, nullable=False)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

class AutoBid(Base):
    __tablename__ = "auto_bids"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    product_id = Column(Integer, nullable=False, index=True)
    user_id = Column(Integer, nullable=False, index=True)
    max_amount = Column(DECIMAL(10, 2), nullable=False)
    increment_amount = Column(DECIMAL(10, 2), default=1.00)
    status = Column(String(20), default="active", comment="active:活跃, paused:暂停, completed:完成, cancelled:取消")
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

class EventProduct(Base):
    __tablename__ = "event_products"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    event_id = Column(Integer, nullable=False, index=True)
    product_id = Column(Integer, nullable=False, index=True)
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime, server_default=func.now())










